c2rust-analyze/src/main.rs

```diff
@@ -449,13 +449,65 @@ fn run(tcx: TyCtxt) {
             assert_eq!(local, l);
         }
 
+        // Find all calls to `malloc`, `calloc`, `realloc`, and `free`
+        // and track their destination locals as being libc::c_void
+        // pointers to special-case downstream casts
+        for bb_data in mir.basic_blocks().iter() {
+            if let Some(term) = &bb_data.terminator {
+                let term: &Terminator = term;
+                if let TerminatorKind::Call {
+                    ref func,
+                    ref args,
+                    destination,
+                    target: _,
+                    ..
+                } = term.kind
+                {
+                    let func_ty = func.ty(&mir.local_decls, tcx);
+                    use crate::util::Callee::*;
+
+                    let assert_libc_cvoid = |p: &Place<'tcx>| {
+                        let deref_ty = p
+                            .ty(acx.local_decls, acx.tcx())
+                            .ty
+                            .builtin_deref(true)
+                            .unwrap()
+                            .ty
+                            .kind();
+                        assert_matches!(deref_ty, TyKind::Adt(adt, _) => {
+                            assert_eq!(tcx.def_path(adt.did()).data[0].to_string(), "ffi");
+                            assert_eq!(tcx.item_name(adt.did()).as_str(), "c_void");
```

Comment:
What about `libc::c_void`?

Also, this works for both `std::` and `core::` `ffi::c_void`?

Reply:
AFAIK this works for `libc::c_void`. That's what `alloc.rs` and `lighttpd-minimal/src/main.rs` are tested with. What changes are you suggesting be made?

Reply:
Hmm, I'm guessing `libc::c_void` now just re-exports `core::ffi::c_void`, so from inside `rustc` we just see that it's `core::ffi::c_void`.  If we want to support older versions, then we'd need to allow `libc::c_void`, but I don't think that's important, so it should be good as is.